### ISSUE TITLE
Restarting Kafka causes consumer offset errors

### ISSUE THREAD

**Issue by dotanrs** [SOHU-Co/kafka-node#1266]
### Environment

* Node version: v8.11.1
* Kafka-node version: 4.0.1
* Kafka version: 2.11-0.10.1.0

### For specific cases also provide
* Number of Brokers: 1
* Number partitions for topic: 1

### The problem

If kafka restarts while the consumer is working, and the consumer doesn't fail, it has very a weird behavior that looks like a bug. This is the flow that I get:

__before crash__
* several messages were sent. latest message received before crash has offset 17, highWaterOffset 18 (according to consumer)
* kafka-consumer-groups --describe: 
 * current offset = 5 (this was the offset when consumer started)
 * log-end-offset = 18
 * lag = 13

__kafka closes__
* consumer prints errors, including 
  * NestedError: refreshBrokerMetadata failed
  * NestedError: topic/partition change check failed

__kafka starts__
* consumer reads all messages from the start (meaning, the 13 messages that were sent after it started)
* this time, the offset of the first message is **22**, highwateroffset is still **18**.
* consumer prints the following log every few milliseconds until stopped:
`replacing ExampleTopic14-0 stale offset of 35 with 0 +113ms`
* later messages append the offset

[Complete debug logs](https://gist.github.com/dotanrs/ca3379bc688cfe2a6824e91fdf77d56c)

the reason this is a problem for me, is that if I commit the offset (which in this example I didn't for simplicity) than the consumer tells Kafka that it's offset is > the `log-end-offset`, which means it has a negative `lag` and this is a whole mess. (details below)

### consumer code
```
const topicsList = [consts.topicName];
const consumerOptions = {
  kafkaHost,
  groupId: consts.consumerGroupName,
  topicsList,
  sessionTimeout: 30000,
  protocol: ['roundrobin'],
  fromOffset: 'latest',
  autoCommit: false,
};

const consumer = new kafka.ConsumerGroup(consumerOptions, topicsList);

consumer.on('message', message => {
  console.log(`received ${JSON.stringify(message)}`);
});

consumer.on('error', e => {
  console.error('Got error in consumer!');
  console.error(e);
});
```

-----

What happens when I do commit the offset

the difference in the code:
```
autoCommit: true
```

in a nutshell:

last message before crash:
kafka: current-offset: 10, log-end-offset: 10, lag: 0
consumer: offset: 9, highwater: 10

after crash:
consumer didn't re-read messages (as expected)
next message: offset: 20, highwater: 11
kafka: current: 21, log-end: 11, lag: -10
debug logs:
` kafka-node:ConsumerGroup replacing ExampleTopic19-0 stale offset of 21 with 0 +7s` times infinity

when another message is received:
consumer reads **both** new message and previous one, kafka lag grows by 1.

[full logs](https://gist.github.com/dotanrs/10a3b443fd917eb7292b19f063a2140c)

**Comment by hyperlink** [comment]
Can you test with the latest 4.1.3? There have been fixes dealing with offsets and compressed topics/messages.

**Comment by taran-1898** [comment]
Any update on this ?

**Comment by kaaljabr** [comment]
I have an issue with this too. When I pause and resume the consumer stream it goes out of sync.
